content/advent-2018/llvm-ir-and-go.md

```diff
@@ -0,0 +1,311 @@
++++
+title = "LLVM IR and Go"
+date = 2017-12-19T08:00:00Z
+author = ["Robin Eklind"]
+series = ["Advent 2018"]
++++
+
+<!-- TODO: remember to update date to 2018 -->
+
+In this post, we'll look at how to build Go programs -- such as compilers and static analysis tools -- that interact with the LLVM compiler framework using the LLVM IR assembly language.
+
+_**TL;DR** we wrote a library for interacting with LLVM IR in pure Go, see links to [code](https://github.com/llir/llvm) and [example projects](https://github.com/llir/llvm#users)._
+
+<!-- TODO: remove 3. Building a toy compiler in Go section? If time allows, add it :) -->
+<!-- 3. Building a toy compiler in Go -->
+
+1. [Quick primer on LLVM IR](#quick-primer-on-llvm-ir)
+2. [LLVM IR library in pure Go](#llvm-ir-library-in-pure-go)
+3. [Closing notes](#closing-notes)
+4. [Further resources](#further-resources)
+
+## Quick primer on LLVM IR
+
+_(For those already familiar with LLVM IR, feel free to [jump to the next section](#llvm-ir-library-in-pure-go))._
+
+[LLVM IR](https://llvm.org/docs/LangRef.html) is a low-level intermediate representation used by the [LLVM compiler framework](http://llvm.org/). You can think of LLVM IR as a platform-independent assembly language with an infinite number of function local registers.
```

Comment:
The way I usually explain IR is by calling it a weird programming language that is convenient for compilers to perform optimizations on. It has (constant) local variables called registers and all control flow (if, while, for) is replaced with goto. The only thing that's really different from other programming languages is the phi node, which is explained further down. Even the `getelementptr` instruction that is usually hard to grasp is just pointer calculation, which most languages that support pointers have special syntax for (like `&buf[3]`).

Reply:
I like your explanation. Will leave it as is for now, but would be good to integrate your explanation when you make the follow-up LLVM post later :)